### DISCUSSION TITLE
Unable to login to Argo UI embedded inside python flask application via iframe. How to fix?

### DISCUSSION THREAD

**Discussion by rakesh-vishwabrahmana** [argoproj/argo-workflows#11837] [Q&A] [answered]
Unable to login to ArgoCD UI which is embedded inside my python flask application. I have changed the configuration of argocd ConfigMap,

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: argocd-cmd-params-cm
    app.kubernetes.io/part-of: argocd
  name: argocd-cmd-params-cm
data:
  server.x.frame.options: allow-from "https://my-app.com/"
  server.content.security.policy: frame-ancestors *
```

So that, I'm able to access the page but unable sign in using admin credentials or using SSO with Microsoft.

But generally I'm able to sign in to the argocd UI using SSO and admin credentials from https://argocd.my-app.com/ without any issues. But I'm unable to sign in from my application which is embedded.

Error message: 400 (Bad Request), http: named cookie not present and data length is less than nonce size

Any advise?

Thanks

**Comment by agilgur5** [comment]
By "embedded" do you mean as an [`iframe`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)? Since you have `frame-ancestors`, that's what it sounded like.

It's been some years since I've used `iframe`s, but IIRC the Cookie is still on the same URL, as it is only being used within the `iframe` and not outside it.

> How do I configure sameSite=None for this? is that works?

Configuring that would [allow for CSRF](https://web.dev/samesite-cookies-explained/#changes-to-the-default-behavior-without-samesite). Even if that were an optional config, it would mean that your server would be susceptible to CSRF now.

You should also be able to workaround this if your Argo Server and Flask application are hosted on [the same domain](https://andrewlock.net/understanding-samesite-cookies/#samesite-strict-cookies). (Cookies are usually [sent in `Strict` mode](https://github.com/search?q=repo%3Aargoproj%2Fargo-workflows%20samesite&type=code))

<hr>

Also back-linking the similar issue that was opened in CD: https://github.com/argoproj/argo-cd/issues/15565

**Reply by rakesh-vishwabrahmana** [reply]
Hi @agilgur5, thanks for the update. 

for better understanding, I just embedded argo workflows and argocd UI to my flask application, as it addresses like, `https://www.my-app.com/argo_workflows` and `https://www.my-app.com/argo_cd`. When I access those pages, it was just shown an error `refused to connect` 

then, I configured the the installation manifest like `server.x.frame.options: allow-from "https://my-app.com/"` and `server.content.security.policy: frame-ancestors *` as I mentioned here [issue](https://github.com/argoproj/argo-cd/issues/15565)

After these changes, it's working and I'm able to access the pages, but unable to login as I raised my query. 

Also, I see the cookies in browser, which shows `SameSite: Strict` for argo workflows and `SameSite: Lax` for argocd. 

As you advised, the [link](https://github.com/search?q=repo%3Aargoproj%2Fargo-workflows%20samesite&type=code) it just shows for few languages, how should I fix this issue with my case? It it possible to configure or do any changes from installation.yaml [ArgoCD](https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/install.yaml)

Thanks

**Reply by agilgur5** [reply]
Could you be more specific about what you mean by "embedded"? Are you using an `iframe`? Are you just proxying via your Flask application? If you could provide some code on the "embed", I would probably be able to decipher pretty quickly.

If it's an `iframe`, hosting on [the same domain](https://andrewlock.net/understanding-samesite-cookies/#samesite-strict-cookies) as I mentioned above should solve it in a secure fashion.

If you're proxying, then it would be better to have your front proxy to do that instead of Flask. For instance, if you have NGINX (the most common k8s Ingress Controller) in front of Flask, you can just have NGINX proxy to the Argo Server directly, which would be best practice and much less error prone. You also probably want to set [`BASE_HREF`](https://argoproj.github.io/argo-workflows/argo-server/#base-href) if you're proxying to a subpath.

> then, I configured the the installation manifest like `server.x.frame.options: allow-from "https://my-app.com/"` and `server.content.security.policy: frame-ancestors *` as I mentioned here [issue](https://github.com/argoproj/argo-cd/issues/15565)

That's only relevant for Argo CD's Server. Did you configure Argo Workflows's Server in a specific way? You said this change made both work, but it is only a change to one of the two.

**Reply by rakesh-vishwabrahmana** [reply]
Yes, I used Iframe in my `apps/argocd.py` file and here is that, 

```
html.Div([
      html.Iframe(
          src='https://argocd.XX.XX.XX.nip.io/',
          style={"height": "1067px", "width": "100%"},
          id = 'embedding',
      ),
```

And, yes, I do configured argo workflow as same, from the `argo-server Deployment` like, 

```
spec:
      containers:
      - args:
        - server
        - --auth-mode=sso
        - --x-frame-options=allow-from "https://my-app.com/"
```
These services I exposed through nginx ingress only, and even my application too.

**Reply by agilgur5** [reply]
Thanks for providing the details, that helps quite a bit.

> `src='https://argocd.XX.XX.XX.nip.io/'`
> `- --x-frame-options=allow-from "https://my-app.com/"`

So if that accurately reflects your set-up, your Flask app and your Argo CD server are on different domains, which would indeed result in a `SameSite` issue. Hosting them on [the same domain](https://andrewlock.net/understanding-samesite-cookies/#samesite-strict-cookies) would fix this.

Also is there a specific reason you're using an `iframe` instead of just proxying the subpath to the Argo Server directly via your ingress? As that would simplify everything.

**Reply by rakesh-vishwabrahmana** [reply]
Thanks for the information. It helps a lot for me. I will try to host them on same domain and check. 

I understood that, [SameSite=None](https://andrewlock.net/understanding-samesite-cookies/#samesite-none-cookies) should work. Please update if there a spec doc to configure or update on hosting different domains to same domain, that would be helpful.

Also, I tried deploying the `argocd server` using ingress as sub path to my application like, `https://my-app.com/argocd`, but the issue is, the service is not able to work from different namespace
 
`/argocd(.*)   argocd-server:80 (<error: endpoints "argocd-server" not found>)` 

because, I'm using `Prod `namespace for my application and using `argo `and `argocd `namespaces for these.

**Reply by rakesh-vishwabrahmana** [reply]
Also, I tried testing on localhost, `src='https://localhost:8080/` but unable to login to argocd UI using admin credentials. 
I see in browser cookies, `argocd.token`, `authorization `and `session` are on same domain `localhost`, but the login functionality not working.

**Reply by rakesh-vishwabrahmana** [reply]
Hi @agilgur5, Could you please update on this. Thank you in advance

**Reply by agilgur5** [reply]
> I understood that, [SameSite=None](https://andrewlock.net/understanding-samesite-cookies/#samesite-none-cookies) should work. Please update if there a spec doc to configure or update on hosting different domains to same domain, that would be helpful.

Yes, it would work, but that would make it insecure as your site would be susceptible to CSRF. So I don't think we should add such an option. `iframe`s have a lot of issues, so I just generally wouldn't recommend using them if you can avoid them. Proxying correctly would be what we should recommend.
I don't think that would fit in the docs as it's not specific to Workflows, that's just an `iframe` issue in general. This is also the first time I've heard someone trying to use an `iframe`, so this isn't a common issue.

> I see in browser cookies, `argocd.token`, `authorization `and `session` are on same domain `localhost`,

Yea port [isn't used](https://web.dev/same-site-same-origin/#same-site-cross-site) for `SameSite`, so if they're all `localhost`, they should work the same.

> but the login functionality not working.

It might be breaking for a different reason, I'm not sure. It may be because you haven't specified `--base-href` / `BASE_HREF`, as I mentioned [above](https://github.com/argoproj/argo-workflows/discussions/11837#discussioncomment-7065404). Would have to see some error logs from your browser and the server at least to try to diagnose that.

Not using `iframes` would definitely simplify things though.

**Reply by agilgur5** [reply, marked as answer]
> because, I'm using `Prod `namespace for my application and using `argo `and `argocd `namespaces for these.

Oh. I can help you with this one. You can create a second `Ingress` (or second set of `paths`) in the `argo` and `argocd` namespaces that use the same `host` as your application.

If you can do that, you should no longer need any `iframes`, which simplifies things quite a bit (`iframes` are a bit of a hacky workaround)

**Reply by rakesh-vishwabrahmana** [reply]
Thank you for the information. Apologies for the late response. As advised I prefer to route using ingress.

**Reply by rakesh-vishwabrahmana** [reply]
Also, just wanted to check, I have installed `nginx-ingress` in `prod` namespace. If I install again `nginx-ingress` in `prod2` namespace, is it will expose another LoadBalancerIP? because current `ingress IP` mapping with my  application with DNS.

**Reply by agilgur5** [reply]
You don't need to install a second Ingress Controller (`nginx-ingress` in your case). You can create [`Ingress`](https://kubernetes.io/docs/concepts/services-networking/ingress/#the-ingress-resource) resources in a different namespaces, and the same Ingress Controller can watch them in any namespace (so long as it is specified in the `ingressClassName` field or is the default ingress class)

**Reply by rakesh-vishwabrahmana** [reply]
That's cool, it works with the `ingressClassName`. Thank you
